python/cuda_parallel/tests/test_reduce.py

```diff
@@ -251,7 +251,6 @@ def test_device_sum_map_mul2_count_it(
     i_input = iterators.TransformIterator(
         mul2,
         iterators.CountingIterator(start_sum_with, value_type=vtn_inp),
-        op_return_value_type=vtn_out,
     )
     _test_device_sum_with_iterator(
         l_varr, start_sum_with, i_input, dtype_inp, dtype_out, use_numpy_array
```

Comment:
Here we're now passing `dtype_out` under the **assumption** that that's what we're getting from `mul2`. What happens if `numba.cuda.compile` infers another type? For example, if numba infers `int64` but users pass `int32` as `dtype_out`? Do we get a runtime error (that would seem good), or will the `int64` be silently truncated?

Reply:
>  For example, if numba infers int64 but users pass int32 as dtype_out? Do we get a runtime error (that would seem good), or will the int64 be silently truncated?

Hmm - good question. I think it comes down to what happens in the call to `thrust::transform` when the unary op returns a different type than the output type. From the [docs](https://en.cppreference.com/w/cpp/algorithm/transform) for `std::transform`:

> The [return type of the unary op] `Ret` must be such that an object of type `OutputIt` can be dereferenced and assigned a value of type `Ret`.​

So, if the return type is different from `dtype_out`, it will be converted to `dtype_out` when being assigned to the output iterator. If it is not convertible, I think that's undefined behaviour - although @gevtushenko, could you please confirm that?

In your specific scenario:

> if numba infers int64 but users pass int32 as dtype_out

Yes, it seems like you will get silent truncation. But that kind of silent truncation should not be unfamiliar, even to Python users:

```python
>>> inp = np.ones(128, dtype="int8")  # the sum of this array should give `128`

>>> out = np.zeros([], dtype="int8")  # int8 is not wide enough to hold the result
>>> print(np.sum(inp, out=out))
array(-128, dtype=int8)

>>> out = np.zeros([], dtype="int32")  # int32 is wide enough to hold the result
>>> print(np.sum(inp, out=out))
array(128, dtype=int32)
```

Reply:
Thanks for the explanation! I agree this is good as is for the purposes of this PR. The undefined behavior worries crossed my mind, too, but I think if it's an actual problem, we'll hit on it while we work on the examples and benchmarks.

Reply:
The type arithmetic in reduction supports the following to be distinct types:

1. Input iterator value type
2. Output iterator value type (output iterators don't have to have it, but let's keep it for discussion)
3. Initial value type
4. Invoke result type of binary operator
5. Accumulator type

Accumulator type is invoke result type of binary operator applied to input iterator value type and initial value type. Accumulator is later written into output iterator. We shouldn't take output iterator value type into account when compiling any of the operators. First of all, when we introduce discard iterator, there won't be any value types. When we introduce transform output iterator, it's value type will be different from what it expects to be assigned. For instance:

```cpp
bool a[3];

auto it = thrust::make_transform_output_iterator(
  a, []__host__ __device__(int val) -> bool { 
    return val > 42 ? false : true; 
});
``` 

here output iterator value type is bool, but iterator expects int.

Reply:
> We shouldn't take output iterator value type into account when compiling any of the operators

Understood, thanks!